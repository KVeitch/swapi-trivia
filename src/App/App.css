* {
box-sizing: border-box;
margin:0;
padding:0;
}

.App {
  background-image: url('../images/tatooinesunset.jpg');
  background-size: cover;
  display: flex;
  flex-direction: column;
  font-family: 'Barlow Condensed', sans-serif;
  letter-spacing: .3em;
  justify-content: space-around;
  height: 100vh;
  text-align: center;
}

.App-logo {
  height: 40vmin;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #09d3ac;
}
