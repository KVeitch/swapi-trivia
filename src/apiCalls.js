export const getMovies = () => {
  return fetch('https://swapi.co/api/films/')
    .then(data => data.json())
    .then(parsedData => parsedData.results)
    .then(movies => 
      movies.sort((movieA, movieB) => movieA.release_date > movieB.release_date ? 1 : -1)
      )
      .then(movies => {
        return movies
      })
}

export const getCharacters = characterUrl => {
  return fetch(characterUrl)
    .then(data => data.json())
    .then(data => {
      return data
    })
}

export const createCharacterList = characters => {
  const characterList = []
  characters.map(character => {
    characterList.push(getCharacters(character))
  })
  return Promise.all(characterList);
}

export const getHomeWorld = homeworld => {
  
}


export const getFilmCharacters = (id) => {
  let url = `https://swapi.co/api/films/${id}`
  return fetch(url).then(res => res.json())
  .then(data => data.characters)
  .then(data => formatData(data))
  .then(chars => getCharacterHomeworld(chars))
  .then(chars => getCharacterSpecies(chars))
  .then(chars => getCharacterFilms(chars))
}

const formatData = (characterUrls) => {
  let charactersData = characterUrls.map(charUrl => {
    return fetch(charUrl).then(res => res.json()).then(data => {
      return {
        films: data.films,
        homeworld: data.homeworld,
        name:data.name,
        species: data.species,
      }
    })
  })
  return Promise.all(charactersData)
}


const getCharacterSpecies = (chars) => {
  let speciesData = chars.map( char => {
    return fetch(char.species).then(res => res.json()).then(species => {
      return {
        ...char,
        species: species.name
        }
      })
    }
  )
  return Promise.all(speciesData)
}

const getCharacterHomeworld = (chars) => {
  let homeworldData = chars.map(char => {
    return fetch(char.homeworld).then(res => res.json()).then(homeworld => {
      return {
        ...char,
        homeworld: homeworld.name,
        population: homeworld.population
      }
    })
  })
  return Promise.all(homeworldData)
}

const getCharacterFilms = (chars) => {
  let films = chars.map(char => {
    let filmTitles = char.films.map(film => {
      return fetch(film).then(res => res.json()).then(film => film.title)
    })
    return Promise.all(filmTitles).then(filmTitles => {
      return {
        ...char,
        films: filmTitles
      }
    }
    )
  })
  return Promise.all(films)
}