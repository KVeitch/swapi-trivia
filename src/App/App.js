import React, { Component } from 'react';
import './App.css';
import Form from '../Form/Form';
import { getMovies, getFilmCharacters } from '../apiCalls';
import MovieContainer from '../MovieContainer/MovieContainer';
import { Switch, Route, Redirect} from 'react-router-dom';
import CharacterContainer from '../CharacterContainer/CharacterContainer'
import UserMenu from '../UserMenu/UserMenu';
import mockFilms from '../mock-data/mockFilm'

class App extends Component {
  constructor() {
    super()
    this.state = {
      user:'',
      userQuote:'',
      userRanking:'',
      movies: [],
      selectedMovie: 1,
      movieCharacters1 : [],
      movieCharacters2 : [],
      movieCharacters3 : [],
      movieCharacters4 : [],
      movieCharacters5 : [],
      movieCharacters6 : [],
      movieCharacters7 : [],
      currentCharacters: [],
      moviesWithImages: []
    }
    console.log('this.state', this.state)
  }
  
  setUser = (user, userQuote, userRanking)=> {
    this.setState({ user, userQuote, userRanking })
  }
  
  setMovies = (mockFilms) => {
    const data = mockFilms.results
    this.setState({
      movies: this.setImages(data)
  })
  }
  
  setCharacters = (mockCharacters) => {
    const characterData = mockCharacters.results
    // getFilmCharacters(this.state.selectedMovie)
    this.setState({ characters : characterData})
  }
  
  // componentDidMount = (mockFilm) => {
    // getMovies(mockFilms)
    // .then(data => this.setState({ movies : data }))
    // getFilmCharacters(this.state.selectedMovie).then(chars => this.setState({ currentCharacters: chars }))
    // .then(()=>console.log('done'))
    // const films = [1,2,3,4,5,6,7]
    // films.forEach(film => {
      // getFilmCharacters(film).then(response => {
        // const movie = `movieCharacters${film}`
        // console.log('movie', movie , response)
        // this.setState({ [movie]: response })
      // })
    // })
  // }

  componentDidMount() {
    this.setMovies(mockFilms)

  }
  setImages = (data) => {
    const dictionary = {
      2: 'https://images-na.ssl-images-amazon.com/images/I/61yWUYWkBhL._SY445_.jpg',
      4: 'https://images-na.ssl-images-amazon.com/images/I/61zAkpvYLqL._SY741_.jpg',
      1: 'https://ae01.alicdn.com/kf/HTB1h5pCNXXXXXXiaXXXq6xXFXXX9.jpg',
      3: 'https://images-na.ssl-images-amazon.com/images/I/61UpAncAQbL._SY679_.jpg',
      6: 'https://images-na.ssl-images-amazon.com/images/I/51UdiBUkerL.jpg',
      5: 'https://images-na.ssl-images-amazon.com/images/I/814Cbv8EftL._SY679_.jpg'
    }
    const moviesWithImages = data.map(movie => {
      const movieImage = dictionary[movie.episode_id];

      return {
        ...movie,
        movieImage
      }
    })
    this.setState({ movies: moviesWithImages })
    console.log('this.state', this.state.movies)
  }

  changeSelectedMovie = (movieId) => {
    this.setState({ selectedMovie : movieId })
  }

  userSignOut = () => {
    this.setState({user: '', userQuote: '', userRanking: ''})
  }

  render() {
    const charProps = `this.props.movieCharacters${this.state.selectedMovie}`;
    return (
      <div className="App">
          <Route exact path='/' render={ (props)=> <Form {...props} setUser={this.setUser} setMovies={this.setMovies} setCharacters={this.setCharacters} setImages={this.setImages}/>} />
        {this.state.movies ? <Route exact path='/movies' render={(props) => <MovieContainer {...props} movies={this.state.movies} changeSelectedMovie={this.changeSelectedMovie} />} /> 
        : ''}
        <Route exact path='/movies' render={(props) => <UserMenu {...props} user={this.state.user} userQuote={this.state.userQuote} userRanking={this.state.userRanking} userSignOut={this.userSignOut} />} />
          <Route exact path={`/movies/${this.state.selectedMovie}`} render={ (props)=> <CharacterContainer {...props} characters={this.state.currentCharacters} />} />
          <Redirect to='/' />
      </div>
    );  
  }
}

export default App;