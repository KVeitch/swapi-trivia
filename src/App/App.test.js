import React from 'react';
import { shallow } from 'enzyme';
import ReactDOM from 'react-dom';
import App from './App';
import { getMovies, getFilmCharacters } from '../apiCalls'
jest.mock('../apiCalls.js')

describe('App', () => {
  let wrapper;

  beforeEach(() => {
    getMovies.mockImplementation(() => {
      return Promise.resolve([{}, {}])
    });

    getFilmCharacters.mockImplementation(() => {
      return Promise.resolve([{}, {}])
    });

    wrapper = shallow(<App />)
  });

  it('should match the snapshot with all the data passed in correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('componentDidMount', () => {
    it('should retrieve movies after mounting', () => {
      shallow(<App />);
      expect(getMovies).toHaveBeenCalled();
    });
  });

  describe('default values', () => {
    it('should have default values of empty strings for user, userQuote, userRanking, and selectedMovie', () => {
      expect(wrapper.state().user).toEqual('');
      expect(wrapper.state().userQuote).toEqual('');
      expect(wrapper.state().userRanking).toEqual('');
      expect(wrapper.state().selectedMovie).toEqual('');
    });

    it('should have a default value of false for isCurrentCharactersLoaded', () => {
      expect(wrapper.state().isCurrentCharactersLoaded).toEqual(false);
    });
  });

  describe('setUser', () => {
    it('should set the state of user, userQuote, userRanking', () => {
      wrapper.setState({ user: 'user', userQuote: 'userQuote', userRanking: 'userRanking' });
    });
  });

  describe('setCurrentCharacters', () => {
    it('should retrieve the characters', async () => {
      await wrapper.instance().setCurrentCharacters();
      expect(getFilmCharacters).toHaveBeenCalled();
    });
  });

  describe('resetIsCurrentCharacterLoaded', () => {
    it('should set the state of isCurrentCharactersLoaded to false and selectedMovie to an empty string', () => {
      wrapper.setState({ isCurrentCharactersLoaded: true, selectedMovie: 'Hello World' });
      wrapper.instance().resetIsCurrentCharacterLoaded();
      expect(wrapper.state().isCurrentCharactersLoaded).toEqual(false);
      expect(wrapper.state().selectedMovie).toEqual('');
    });
  });

  describe('changeSelectedMovie', () => {
    it('should call setCurrentCharacters', async () => {
      await wrapper.instance().changeSelectedMovie();
      expect(wrapper.instance().setCurrentCharacters).toHaveBeenCalled();
    });

    it.skip('should update state.selectedMovie to the movie number passed in', async () => {
      await wrapper.instance().changeSelectedMovie(2)
      expect(wrapper.state().selectedMovie).toEqual(2)
    });
  });

  describe('setImages', () => {
    it('should update state.movies to update with images', () => {
      wrapper.instance().setImages();
      expect(wrapper.state().movies).toEqual([{}, {}]);
    });
  });

  describe('signUserOut', () => {
    it('should reset the state when a user signs out', () => {
      wrapper.setState({
        user: 'Young Julian',
        userQuote: 'Hihihihi',
        userRanking: 'Young Padawan',
        movies: ['', ''],
        selectedMovie: 'A New Hope',
        currentCharacters: ['', ''],
        isCurrentCharactersLoaded: true,
        favoriteCharacters: ['', '']
      });
      wrapper.instance().signUserOut();
      expect(wrapper.state()).toEqual({
        user: '',
        userQuote: '',
        userRanking: '',
        movies: [],
        selectedMovie: '',
        currentCharacters: [],
        isCurrentCharactersLoaded: false,
        favoriteCharacters: []
      })
    });
  });

});
