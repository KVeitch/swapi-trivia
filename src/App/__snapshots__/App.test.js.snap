// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot with all the data passed in correctly 1`] = `
<div
  className="App"
>
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/movies"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/movies/:id"
    render={[Function]}
  />
  <Route
    path="/movies"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/favorites"
    render={[Function]}
  />
</div>
`;
